### ISSUE TITLE
chef cronjob disabled in some VMs 

### ISSUE THREAD

**Issue by rubeon** [chef-partners/azure-chef-extension#147]
Hi,

it looks like there was some issue last night  with an attempted update to `1210.12.104.1000` that disabled chef in some of our VMs. 

**waagent.log**

```
2016/06/21 01:38:10 Retrieved GoalState from Azure Fabric.
2016/06/21 01:38:10 ExpectedState: Started
2016/06/21 01:38:10 ContainerId: [snip]
2016/06/21 01:38:10 RoleInstanceId: [snip]
2016/06/21 01:38:10 Private key with thumbprint: [snip] was retrieved.
2016/06/21 01:38:10 Public cert with thumbprint:  [snip] was retrieved.
2016/06/21 01:38:10 Posted Role Properties. CertificateThumbprint= [snip]
2016/06/21 01:38:13 Found Plugin: Chef.Bootstrap.WindowsAzure.LinuxChefClient version: 1210.12.104.1000
2016/06/21 01:38:13 Downloading plugin manifest: Chef.Bootstrap.WindowsAzure.LinuxChefClient from  [snip]
2016/06/21 01:38:13 Plugin server is:  [snip]
2016/06/21 01:38:13 Bundle URI =  [snip]
2016/06/21 01:38:13 Plugin bundle [snip] successfully length = 75695
2016/06/21 01:38:13 Extracted  [snip]/Chef.Bootstrap.WindowsAzure__LinuxChefClient__1210.12.104.1000 to /var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:13 Found RuntimeSettings for Chef.Bootstrap.WindowsAzure.LinuxChefClient V 1210.12.104.1000
2016/06/21 01:38:13 SetHandlerState: Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000, NotInstalled
2016/06/21 01:38:13 Spawned disable.sh PID 34936
2016/06/21 01:38:18 disableCommand completed.
2016/06/21 01:38:18 SetHandlerState: Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000, Disabled
2016/06/21 01:38:18 Chef.Bootstrap.WindowsAzure.LinuxChefClient version 1210.12.103.1000 is disabled
2016/06/21 01:38:18 Copy status file from old plugin dir to new
2016/06/21 01:38:18 Spawned update.sh PID 35166
2016/06/21 01:38:23 ERROR:Process 35166 returned non-zero exit code(1)
2016/06/21 01:38:23 ERROR:Update failed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:23 Spawned uninstall.sh PID 35201
2016/06/21 01:38:28 uninstallCommand completed.
2016/06/21 01:38:28 SetHandlerState: Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000, NotInstalled
2016/06/21 01:38:28 Uninstall completeChef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000
2016/06/21 01:38:28 Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000 extension files deleted.
2016/06/21 01:38:28 Found RuntimeSettings for Chef.Bootstrap.WindowsAzure.LinuxChefClient V 1210.12.104.1000
2016/06/21 01:38:28 Spawned install.sh PID 35221
2016/06/21 01:38:33 installCommand completed.
2016/06/21 01:38:33 SetHandlerState: Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000, Installed
2016/06/21 01:38:33 Installation completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:33 Spawned enable.sh PID 35241
2016/06/21 01:38:38 enableCommand completed.
2016/06/21 01:38:38 SetHandlerState: Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000, Enabled
2016/06/21 01:38:38 Enable completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:38 Processing completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:38 Found Plugin: Microsoft.OSTCExtensions.LinuxDiagnostic version: 2.3.9001
2016/06/21 01:38:38 Found RuntimeSettings for Microsoft.OSTCExtensions.LinuxDiagnostic V 2.3.9001
2016/06/21 01:38:38 Spawned diagnostic.py -enable PID 35258
2016/06/21 01:38:39 LinuxAzureDiagnostic started to handle.
2016/06/21 01:38:39 [Microsoft.OSTCExtensions.LinuxDiagnostic-2.3.9001] cwd is /var/lib/waagent/Microsoft.OSTCExtensions.LinuxDiagnostic-2.3.9001
2016/06/21 01:38:39 [Microsoft.OSTCExtensions.LinuxDiagnostic-2.3.9001] Change log file to /var/log/azure/Microsoft.OSTCExtensions.LinuxDiagnostic/2.3.9001/extension.log
2016/06/21 01:38:43 enableCommand completed.

```

And the extension log:
**/var/log/azure/Chef.Bootstrap.WindowsAzure.LinuxChefClient/1210.12.104.1000/CommandExecution.log**

```
2016/06/21 01:38:13 Downloading plugin manifest: Chef.Bootstrap.WindowsAzure.LinuxChefClient from [snip]/Chef.Bootstrap.WindowsAzure_LinuxChefClient_europenorth_manifest.xml
2016/06/21 01:38:13 Plugin server is: rdfepirv2db3prdstr02.blob.core.windows.net
2016/06/21 01:38:13 Bundle URI = [snip]/Chef.Bootstrap.WindowsAzure__LinuxChefClient__1210.12.104.1000
2016/06/21 01:38:13 Plugin bundle [snip]/Chef.Bootstrap.WindowsAzure__LinuxChefClient__1210.12.104.1000downloaded successfully length = 75695
2016/06/21 01:38:13 Extracted [snip]/Chef.Bootstrap.WindowsAzure__LinuxChefClient__1210.12.104.1000 to /var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:13 Found RuntimeSettings for Chef.Bootstrap.WindowsAzure.LinuxChefClient V 1210.12.104.1000
2016/06/21 01:38:13 Spawned disable.sh PID 34936
2016/06/21 01:38:18 disableCommand completed.
2016/06/21 01:38:18 Chef.Bootstrap.WindowsAzure.LinuxChefClient version 1210.12.103.1000 is disabled
2016/06/21 01:38:18 Spawned update.sh PID 35166
2016/06/21 01:38:23 Process 35166 returned non-zero exit code(1)
2016/06/21 01:38:23 Update failed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:23 Spawned uninstall.sh PID 35201
2016/06/21 01:38:28 uninstallCommand completed.
2016/06/21 01:38:28 Uninstall completeChef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000
2016/06/21 01:38:28 Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000 extension files deleted.
2016/06/21 01:38:28 Found RuntimeSettings for Chef.Bootstrap.WindowsAzure.LinuxChefClient V 1210.12.104.1000
2016/06/21 01:38:28 Spawned install.sh PID 35221
2016/06/21 01:38:33 installCommand completed.
2016/06/21 01:38:33 Installation completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:33 Spawned enable.sh PID 35241
2016/06/21 01:38:38 enableCommand completed.
2016/06/21 01:38:38 Enable completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
2016/06/21 01:38:38 Processing completed for Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000

```

This left the system in a state where the chef-client is no longer being run in the crontab, and hasn't checked in with the Chef server since that event.

There were several other machines that didn't try to update, but I'm not sure how they might be configured differently.  Is there a way to block this kind of update from happening?

**Comment by NimishaS** [comment]
@rubeon , can you please share your `/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000/config/0.settings`, ` /var/log/azure/custom.log` and `/var/lib/waagent/ExtensionsConfig.1.xml` file?

Also please share how you created the VM with extension.

**Comment by rubeon** [comment]
Hi,

0.settings and ExtensionsConfig.1.xml have a lot of sensitive data, so I'm not going to share it here.  What in particular would you like to see from those files?

**/var/log/azure/custom.log**
```
/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.102.1000
Call for Checking linux distributor
Starting installation process:
11:43:56
Checking for curl...
Detected curl...
Reading chef-client version from settings file
Installing chef-client package
Preparing...                          ########################################
Updating / installing...
chef-12.9.41-1.el7                    ########################################
Thank you for installing Chef!
[Wed May  4 11:44:09 UTC 2016] Package Chef installed successfully.
End of installation:
11:44:09
[Wed May  4 11:44:09 UTC 2016] Installing Azure Chef Extension gem
Successfully installed azure-chef-extension-0.0.1
1 gem installed
[Wed May  4 11:44:11 UTC 2016] Azure Chef Extension gem installation succeeded
/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.102.1000
/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000
[Sat May 28 00:35:30 UTC 2016] Not doing install, as auto update is false
/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.103.1000
/var/lib/waagent/Chef.Bootstrap.WindowsAzure.LinuxChefClient-1210.12.104.1000
[Tue Jun 21 01:38:28 IST 2016] Reading environment variables
[Tue Jun 21 01:38:28 IST 2016] No environment variables found
[Tue Jun 21 01:38:28 IST 2016] Not doing install, as auto update is false
```
The extension was added via Azure Resource Manager Template.  The snippet is here:

```json
// variables
    "extensionProperties": {
      "protectedSettings": {
        "validation_key": "[parameters('chefValidationKey')]"
      },
      "publisher": "Chef.Bootstrap.WindowsAzure",
      "typeHandlerVersion": "1210.12",
      "type": "LinuxChefClient",
      "settings": {
        "bootstrap_options": {
          "environment": "[parameters('chefEnvironment')]",
          "secret": "[parameters('chefSecret')]",
          "validation_client_name": "[parameters('chefValidationClientName')]",
          "chef_node_name": "[parameters('chefNodeName')]",
          "chef_server_url": "[parameters('chefUrl')]"
        },
        "runlist": "role[default]",
        "client_rb": "[parameters('chefClientRbOptions')]"
      }
    },

```

```json

// resources

    {
      "name": "[concat(variables('machineName'), '/LinuxChef')]",
      "dependsOn": [
        "[concat('Microsoft.Compute/virtualMachines/',concat(variables('machineName')))]"
      ],
      "type": "Microsoft.Compute/virtualMachines/extensions",
      "properties": "[variables('extensionProperties')]",
      "apiVersion": "[variables('apiVersion')]",
      "location": "[variables('location')]"
    }

```

**Comment by NimishaS** [comment]
@rubeon , please share the value of `autoUpdateClient` from `0.settings` file and `autoUpgrade` from `ExtensionsConfig.1.xml`.

**Comment by rubeon** [comment]
The extension config for the Chef client has the following:
```xml
<Plugin name="Chef.Bootstrap.WindowsAzure.LinuxChefClient" version="1210.12.103.1000" ... autoUpgrade="false">

```

There is nothing about updates in 0.settings.

Thanks

**Comment by NimishaS** [comment]
Thanks for sharing the information @rubeon.
The default value of `autoUpdateClient` is false if it's not specified in the `0.settings` file. This issue is not related to the latest release. It has been a known issue for some time.

Setting `autoUpdateClient = true` in the template while VM provisioning will fix this issue. Though this solution will not help on the existing VMs that you have already created.

**Comment by NimishaS** [comment]
@rubeon , we have published extension `1210.12.105.1000` where we removed some flags including `autoUpdateClient` to avoid this kind of issue. You won't be facing this again.
